x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/enrich_signal_threat_matches.ts

```diff
@@ -0,0 +1,108 @@
+    const signalId = getSignalId(signalHit);
+    if (!acc.has(signalId)) {
+      acc.set(signalId, signalHit);
+    } else {
+      const existingSignalHit = acc.get(signalId) as SignalSourceHit;
+      const existingQueries = existingSignalHit?.matched_queries ?? [];
+      const newQueries = signalHit.matched_queries ?? [];
+      existingSignalHit.matched_queries = [...existingQueries, ...newQueries];
+
+      acc.set(signalId, existingSignalHit);
+    }
+
+    return acc;
+  }, new Map());
+  const dedupedHits = Array.from(dedupedHitsMap.values());
+  return dedupedHits;
+};
+
+export const buildMatchedIndicator = ({
+  queries,
+  threats,
+  indicatorPath = DEFAULT_INDICATOR_PATH,
+}: {
+  queries: ThreatMatchNamedQuery[];
+  threats: ThreatListItem[];
+  indicatorPath?: string;
+}): ThreatIndicator[] =>
+  queries.map((query) => {
+    const matchedThreat = threats.find((threat) => threat._id === query.id);
+    const indicatorValue = get(matchedThreat?._source, indicatorPath) as unknown;
+    const indicator = [indicatorValue].flat()[0] ?? {};
+    if (!isObject(indicator)) {
+      throw new Error(`Expected indicator field to be an object, but found: ${indicator}`);
+    }
+    const atomic = get(matchedThreat?._source, query.value) as unknown;
+    const type = get(indicator, 'type') as unknown;
+
+    return {
+      ...indicator,
+      matched: { atomic, field: query.field, type },
+    };
+  });
+
+export const enrichSignalThreatMatches = async (
+  signals: SignalSearchResponse,
+  getMatchedThreats: GetMatchedThreats
+): Promise<SignalSearchResponse> => {
+  const signalHits = signals.hits.hits;
+  if (signalHits.length === 0) {
+    return signals;
+  }
+
+  // TODO update hits total to account for deduping
+  const uniqueHits = groupAndMergeSignalMatches(signalHits);
+  const signalMatches = uniqueHits.map((signalHit) => extractNamedQueries(signalHit));
+  const matchedThreatIds = [...new Set(signalMatches.flat().map(({ id }) => id))];
+  const matchedThreats = await getMatchedThreats(matchedThreatIds);
+  const matchedIndicators = signalMatches.map((queries) =>
+    buildMatchedIndicator({ queries, threats: matchedThreats })
+  );
+
+  const enrichedSignals = uniqueHits.map((signalHit, i) => {
+    const threat = get(signalHit._source, 'threat') ?? {};
+    if (!isObject(threat)) {
+      throw new Error(`Expected threat field to be an object, but found: ${threat}`);
+    }
+    const existingIndicatorValue = get(signalHit._source, 'threat.indicator') ?? [];
+    const existingIndicators = [existingIndicatorValue].flat(); // ensure indicators is an array
+
+    return {
+      ...signalHit,
+      _source: {
+        ...signalHit._source,
+        threat: {
+          ...threat,
+          indicator: [...existingIndicators, ...matchedIndicators[i]],
+        },
+      },
+    };
+  });
+  // eslint-disable-next-line require-atomic-updates
```

Comment:
It seems like this linter rule is triggered due to there being an `await` before this assignment happens. I don't see any way that `signals` could be modified in the meantime, but I'd appreciate a quick sanity check.